source/agora/common/Deserializer.d

```diff
@@ -144,95 +144,103 @@ public T deserializeFull (T) (scope ubyte[] data) @safe
 }
 
 /// Ditto
-public T deserializeFull (T) (scope DeserializeDg dg) @safe
+public void deserializePart (T) (ref T record, scope DeserializeDg dg) @safe
 {
-    T value;
-    deserializePart(value, dg);
-    return value;
+    record = deserializeFull!T(dg);
 }
 
 /// Ditto
-public void deserializePart (T) (
-    ref T record, scope DeserializeDg dg,
-    CompactMode compact = CompactMode.Yes)
+public T deserializeFull (T) (
+    scope DeserializeDg dg, CompactMode compact = CompactMode.Yes)
     @safe
 {
     import geod24.bitblob;
 
     // Custom deserialization trumps everything
     static if (hasDeserializeMethod!T)
-        record.deserialize(dg);
+    {
+        T retval = T.init;
+        retval.deserialize(dg);
```

Comment:
Aha, so this part will also need fixing later, like mentioned in the PR comment. Making it a static method so it can return immutable.

Reply:
The signature will be:
```D
struct Foo
{
    static T fromBinary(T) (scope DeserializationDg dg) @safe
     {
         // Magic
     }
}
```

Unfortunately the template is necessary because there's no way to call different function based on whether `T` is `const`, `immutable`, or not.